server/src/main/java/org/elasticsearch/search/crossproject/CrossProjectIndexResolutionValidator.java

```diff
@@ -138,33 +167,82 @@ public static ElasticsearchException validate(
                 // checking if flat expression matched remotely
                 for (String remoteExpression : remoteExpressions) {
                     String[] splitResource = splitQualifiedResource(remoteExpression);
-                    ElasticsearchException exception = checkSingleRemoteExpression(
+                    var projectAlias = splitResource[0];
+                    var resource = splitResource[1];
+
+                    ElasticsearchException remoteException = checkSingleRemoteExpression(
                         remoteResolvedExpressions,
-                        splitResource[0], // projectAlias
-                        splitResource[1], // resource
+                        projectAlias,
+                        resource,
                         remoteExpression,
                         indicesOptions
                     );
-                    if (exception == null) {
+                    if (remoteException == null) {
                         // found flat expression somewhere
                         foundFlat = true;
                         break;
                     }
-                    if (false == isUnauthorized && exception instanceof ElasticsearchSecurityException) {
+                    if (false == isUnauthorized && remoteException instanceof ElasticsearchSecurityException) {
                         isUnauthorized = true;
+                        if (remoteAuthorizationExceptions == null) {
+                            remoteAuthorizationExceptions = new HashMap<>();
+                            remoteUnauthorizedIndices = new HashMap<>();
+                        }
+                        remoteAuthorizationExceptions.putIfAbsent(projectAlias, remoteException);
+                        remoteUnauthorizedIndices.compute(projectAlias, (k, v) -> {
+                            if (v == null) {
+                                v = new ArrayList<>();
+                            }
+                            v.add(resource);
+                            return v;
+                        });
```

Comment:
I think we can extract this as a method and use it here and in the above `isQualifiedExpression` branch?